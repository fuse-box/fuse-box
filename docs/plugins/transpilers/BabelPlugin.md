# BabelPlugin

## Description
The babel plugin is used to transpile code to different dialects of javascript.

## Install
This package depends on the `babel-core` module.

```bash
yarn add babel-core babel-preset-es2015 babel-plugin-transform-react-jsx --dev
npm install babel-core babel-preset-es2015 babel-plugin-transform-react-jsx --save-dev
```

##### Parameters

| Name | Type | Description | ?Default |
| ---- | ---- | ----------- | -------- |
| config | `Object`  | when using other fuse-box only properties, babel config is passed in as .config |  |
| limit2project | `boolean`  | to use this plugin across an entire project (including other modules like npm) | `true` |
| extensions | `Array<string>`  | file extensions to allow with fuse-box | `[".jsx"]`
| test | `Regex`  | files to match | <code>/\\.(j&#124;t)s(x)?$/</code> |




## Examples

### JSX

```js
const {BabelPlugin} = require("fuse-box");

plugins: [
    BabelPlugin({
      config: {
        sourceMaps: true,
        presets: ["es2015"],
        plugins: [
            ["transform-react-jsx"],
        ],
      },
    })
]
```

### Shorthand

when only passing in a babel config, there is no need for .config

```js
plugins: [
  BabelPlugin({
    presets: ["es2015"]
  })
]
```
