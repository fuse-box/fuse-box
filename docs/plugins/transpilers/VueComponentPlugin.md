# VueComponentPlugin

## Description
The VueComponentPlugin is used to transpile `.vue` files into Javascript.

## Install
This plugin depends on the following node modules:
- `vue-template-compiler`
- `vue-template-es2015-compiler`
- `vue-hot-reload-api`
- `postcss-selector-parser`
- `vue`

```bash
# Using yarn:
yarn add vue-template-compiler vue-template-es2015-compiler vue-hot-reload-api vue postcss-selector-parser --dev

# Using npm:
npm install vue-template-compiler vue-template-es2015-compiler vue-hot-reload-api vue postcss-selector-parser --save-dev
```

For more information about the `*.vue` file specifications please see the official [documentation](https://vue-loader.vuejs.org/en/start/spec.html)

## Usage
Just import from FuseBox:
```js
const { VueComponentPlugin } = require('fuse-box')
```

Inject into a chain:
```js
fuse.plugin(
  VueComponentPlugin()
)
```

Or add it to the main config to make it available across bundles:
```js
FuseBox.init({
  plugins: [
    VueComponentPlugin()
  ]
});
```

## Configuration
### Defaults
If no `lang` attributes are specified (see below) the `VueComponentPlugin` will use the following configuration out of the box:

- `<template>` Defaults to `html` and uses [HTMLPlugin](/plugins/html-plugin)
- `<script>` Defaults to `ts` and uses FuseBox's integrated Typescript compiler
- `<style>` Defaults to `css` and uses [CSSPlugin](/plugins/css-plugin)

### Language Attributes
When `VueComponentPlugin` detects a `lang` attribute on a block it will attempt to match with the corresponding FuseBox plugin. For example:

- `<script lang="coffee">` - [CoffeeScriptPlugin](/plugins/coffeescript-plugin)
- `<script lang="js">` - [BabelPlugin](/plugins/babelplugin) (will read configuration from `.babelrc` file)
- `<style lang="less">` - [LESSPlugin](/plugins/less-plugin)
- `<style lang="scss">` - [SASSPlugin](/plugins/sass-plugin)

note: If using lang="js" and configuration item useTypescriptCompiler: true then FuseBox will use the internal Typescript compiler and NOT BabelPlugin


### Using Custom Plugin Chains
If the above functionality doesn't fit your needs, you can override the pre-processing by optionally setting the `script`, `style` or `template` options. This follows the standard FuseBox way of defining plugin chains:

```js
const { FuseBox, VueComponentPlugin, BabelPlugin, SassPlugin, CSSResourcePlugin, PostCSSPlugin, CSSPlugin } = require('fuse-box')

const fsbx = FuseBox.init({
    homeDir: './src',
    output: 'dist/app.js',
    plugins: [
      VueComponentPlugin({
          script: BabelPlugin({ ... }),
          template: HTMLPlugin({ ... })
          style: [
              SassPlugin({ ... }),
              CSSResourcePlugin({ ... }),
              PostCSSPlugin( ... ),
              CSSPlugin({ ... })
          ]
      })
    ]
})
```

note: Overriding a plugin chain for a .vue block will make the VueComponentPlugin ignore any lang attributes.

### External Files
The `VueComponentPlugin` fully understands the `src` attribute and will handle external files just the same as inline content.

### Scoped Styling
Scoped styling is fully supported by using the `scoped` attribute. Support for the `module` attribute will be coming soon.

### HMR
Hot Module Reloading is also fully supported, just enable it on your bundles by calling `.hmr()`. See [here](/page/development#hot-module-reload) for more info.
