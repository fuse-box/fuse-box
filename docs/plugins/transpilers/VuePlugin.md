# VuePlugin

## Description
The Vue plugin is used to transpile .vue files into Javascript.

## Install
This package depends on the `vue-template-compiler`, `vue-template-es2015-compiler` and one of `typescript` and `babel-core` modules.

Also, if you want to include Vue library into the bundle, remember to install `vue` module too.

```bash
# Using yarn:
yarn add vue-template-compiler vue-template-es2015-compiler typescript babel-core vue --dev

# Using npm:
npm install vue-template-compiler vue-template-es2015-compiler typescript babel-core vue --save-dev
```

## Example

```js
// fuse.js

const { FuseBox, VuePlugin } = require('fuse-box')

const fsbx = FuseBox.init({
    homeDir: './src',
    output: 'dist/app.js',
    plugins: [
        VuePlugin()
    ]
})

fsbx.dev()

fsbx.bundle('app.js')
    .instructions('>index.ts')
    .watch()

fsbx.run()

```

```typescript
// src/index.ts

import * as Vue from 'vue/dist/vue.js'
import * as App from './components/App.vue'

new Vue({
    el: '#app',
    render: h => h(App)
})

```

```html
<!-- src/components/App.vue -->

<template>
    <div>
        <p>{{ msg }}</p>
        <input type="text" v-model="msg" />
    </div>
</template>

<script>
    export default {
        name: 'app',
        data () {
            return {
                msg: 'Welcome to Your Vue.js App'
            }
        }
    }
</script>
```

## Script transpiler

By default `VuePlugin` uses `typescript` to compile contents of `<script></script>`. By defining the `lang` attribute of the `<script>` tag, content can be compiled using desired transpiler. Currently only `ts` and `babel` are supported.

### Babel

If `babel` attribute is given, VuePlugin will load babel config from there, otherwise VuePlugin will load `.babelrc` from the root of the project. If no config given and no rc found, it uses default config. Packages `babel-core` and `babel-plugin-transform-es2015-modules-commonjs` must be installed to use babel.

```js
// Configuring babel in fuse.js (Default)
VuePlugin({
    babel: {
        config: {
            "plugins": ["transform-es2015-modules-commonjs"]
        }
    }
})

// .babelrc
{
 "plugins": ["transform-es2015-modules-commonjs"]
}
```

## Template engine preprocessor

By defining the `lang` attribute of the `<template>` tag, content can be compiled to HTML using your favorite template engine.

The `consolidate` module must be installed in addition to the template engine module you want to use.

View a list of all [supported template engines](https://github.com/tj/consolidate.js#supported-template-engines).

### Example using Pug:

```bash
# Using yarn:
yarn add consolidate pug --dev

# Using npm:
npm install consolidate pug --save-dev
```

```html
<!-- src/components/App.vue -->

<template lang="pug">
    .content
        p {{ msg }}
        input(type="text", v-model="msg")
</template>

<script>
    export default {
        data () {
            return {
                msg: 'Welcome to Your Vue.js App'
            }
        }
    }
</script>
```
