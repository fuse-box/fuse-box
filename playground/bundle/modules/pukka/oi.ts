import * as hey from './hey';
import * as oo from './hey';
console.log(hey, oo);
