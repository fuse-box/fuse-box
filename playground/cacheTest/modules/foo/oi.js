export function oi() {
  return 'I am additional oi';
}
