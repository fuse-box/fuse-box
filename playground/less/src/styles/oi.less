body {
  padding: 50px !important;
}
