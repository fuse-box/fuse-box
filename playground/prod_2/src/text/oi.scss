body {
  color: red;
  background-image: url('../assets/logo.png');
}
