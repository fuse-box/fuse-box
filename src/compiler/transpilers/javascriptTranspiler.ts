import { ASTNode } from '../interfaces/AST';
import { ImportType } from '../interfaces/ImportType';
import { ITransformerRequireStatementCollection } from '../interfaces/ITransformerRequireStatements';
import { ITranspiler } from '../interfaces/ITranspiler';
import { createGlobalContext } from '../program/GlobalContext';
import { ITransformerList, transpileModule } from '../program/transpileModule';
import { BrowserProcessTransformer } from '../transformers/bundle/BrowserProcessTransformer';
import { BundleFastConditionUnwrapper } from '../transformers/bundle/BundleFastConditionTransformer';
import { BundlePolyfillTransformer } from '../transformers/bundle/BundlePolyfillTransformer';
import { RequireStatementInterceptor } from '../transformers/bundle/RequireStatementInterceptor';
import { GlobalContextTransformer } from '../transformers/GlobalContextTransformer';
import { DynamicImportTransformer } from '../transformers/shared/DynamicImportTransformer';
import { ExportTransformer } from '../transformers/shared/ExportTransformer';
import { ImportTransformer } from '../transformers/shared/ImportTransformer';
import { JSXTransformer } from '../transformers/shared/JSXTransformer';
import { ITransformerResult } from '../interfaces/ITranformerResult';

export function javascriptTranspiler(props: ITranspiler): ITransformerResult {
  const requireStatementCollection: ITransformerRequireStatementCollection = [];
  function onRequireCallExpression(importType: ImportType, statement: ASTNode) {
    // making sure we have haven't emitted the same property twice
    if (!statement['emitted']) {
      Object.defineProperty(statement, 'emitted', { enumerable: false, value: true });
      requireStatementCollection.push({ importType, statement });
    }
  }

  const defaultTransformers: ITransformerList = [
    GlobalContextTransformer(),
    BundleFastConditionUnwrapper({
      env: props.env,
      isBrowser: props.isBrowser,
      isServer: props.isServer,
    }),
    RequireStatementInterceptor({ onRequireCallExpression }),
    props.target === 'browser' && BrowserProcessTransformer({ env: props.env, onRequireCallExpression }),
    BundlePolyfillTransformer({ ...props, onRequireCallExpression }),

    DynamicImportTransformer({ onRequireCallExpression }),

    JSXTransformer(),

    ImportTransformer({ onRequireCallExpression }),
    ExportTransformer({ onRequireCallExpression }),
  ];
  if (props.transformers && props.transformers.length) {
    const opts = { module: props.module, onRequireCallExpression };
    for (const transformer of props.transformers) {
      defaultTransformers.unshift(transformer(opts));
    }
  }

  transpileModule({
    ast: props.ast as ASTNode,
    globalContext: createGlobalContext({}),
    transformers: defaultTransformers,
  });

  return { ast: props.ast, requireStatementCollection };
}
