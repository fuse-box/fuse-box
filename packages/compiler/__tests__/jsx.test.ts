import { testTranspile } from '../transpilers/testTranpiler';

describe('JSX', () => {
  it('should parse simple jsx', () => {
    const result = testTranspile({
      code: `
        function test(){
            return <div></div>
        }
          `,
    });
    expect(result.code).toMatchSnapshot();
  });

  it('should add text', () => {
    const result = testTranspile({
      code: `
        function test(){
            return <div>1</div>
        }
          `,
    });
    expect(result.code).toMatchSnapshot();
  });

  it('should add just attributes', () => {
    const result = testTranspile({
      code: `
        import oi from "./oi";
        function test(){
          return (<i id="1" f={oi} ></i>)
        }
          `,
    });
    expect(result.code).toMatchSnapshot();
  });

  it('should add just attributes and spread', () => {
    const result = testTranspile({
      code: `
        import oi from "./oi";
        function test(){
          return (<i id="1" f={oi} {...props} ></i>)
        }
          `,
    });

    expect(result.code).toMatchSnapshot();
  });

  it('should have only one spread props', () => {
    const result = testTranspile({
      code: `
        import oi from "./oi";
        function test(){
          return (<i {...props} ></i>)
        }
          `,
    });
    expect(result.code).toMatchSnapshot();
  });

  it('should keep the order', () => {
    const result = testTranspile({
      code: `
        import oi from "./oi";
        function test(){
          return (<i id={1} f={oi} {...props} s={2} {...rest}></i>)
        }
          `,
    });

    expect(result.code).toMatchSnapshot();
  });

  it('should remove empty expressions', () => {
    const result = testTranspile({
      code: `
        import oi from "./oi";
        function test(){
          return (<i> {/* <i></i> */} </i>)
        }
          `,
    });

    expect(result.code).toMatchSnapshot();
  });

  it('should handle JSX fragment', () => {
    const result = testTranspile({
      code: `
        import React from "react";
        function test(){
          return (
            <div>
              <>
                <div>1</div>
              </>
            </div>
          )
        }
          `,
    });

    expect(result.code).toMatchSnapshot();
  });

  it('should work with JSXSpreadChild', () => {
    const result = testTranspile({
      code: `

        function test(){
          return (<i>{...children}</i>)
        }
          `,
    });
    expect(result.code).toMatchSnapshot();
  });
});
