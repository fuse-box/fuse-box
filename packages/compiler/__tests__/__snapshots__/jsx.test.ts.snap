// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSX should add just attributes 1`] = `
"var oi_1 = require(\\"./oi\\");
function test() {
  return React.createElement(\\"i\\", {
    id: \\"1\\",
    f: oi_1.default
  });
}
"
`;

exports[`JSX should add just attributes and spread 1`] = `
"var oi_1 = require(\\"./oi\\");
function test() {
  return React.createElement(\\"i\\", Object.assign({
    id: \\"1\\",
    f: oi_1.default
  }, props));
}
"
`;

exports[`JSX should add text 1`] = `
"function test() {
  return React.createElement(\\"div\\", null, \\"1\\");
}
"
`;

exports[`JSX should handle JSX fragment 1`] = `
"var react_1 = require(\\"react\\");
function test() {
  return react_1.default.createElement(\\"div\\", null, react_1.default.createElement(react_1.default.Fragment, null, react_1.default.createElement(\\"div\\", null, \\"1\\")));
}
"
`;

exports[`JSX should have only one spread props 1`] = `
"function test() {
  return React.createElement(\\"i\\", Object.assign({}, props));
}
"
`;

exports[`JSX should keep the order 1`] = `
"var oi_1 = require(\\"./oi\\");
function test() {
  return React.createElement(\\"i\\", Object.assign({
    id: 1,
    f: oi_1.default
  }, props, {
    s: 2
  }, rest));
}
"
`;

exports[`JSX should parse simple jsx 1`] = `
"function test() {
  return React.createElement(\\"div\\", null);
}
"
`;

exports[`JSX should remove empty expressions 1`] = `
"function test() {
  return React.createElement(\\"i\\", null, \\" \\", \\" \\");
}
"
`;

exports[`JSX should work with JSXSpreadChild 1`] = `
"function test() {
  return React.createElement(\\"i\\", null, children);
}
"
`;
